docs/design/structs-interfaces.md

```diff
@@ -0,0 +1,251 @@
+  Update func(context.Context, ResourceUpdateRequest, ResourceUpdateResponse)
+  Delete func(context.Context, ResourceDeleteRequest, ResourceDeleteResponse)
+  Import func(context.Context, ResourceImportRequest, ResourceImportResponse)
+}
+
+type Importer interface {
+  Import(context.Context, string) ([]*tfprotov5.ImportedResource, []*tfprotov5.Diagnostic)
+}
+
+type Validater interface {
+  Validate(context.Context, tftypes.Value) []*tfprotov5.Diagnostic
+}
+
+type CustomPlanner interface {
+  ModifyPlan(ctx context.Context, priorState, proposedState, config tftypes.Value) (tftypes.Value, []*tfprotov5.Diagnostic)
+}
+```
+
+### Alternative 3: Mix of structs and interfaces
+
+Example framework implementation:
+
+```go
+type Provider struct {
+	Resources     map[string]Resource
+	ConfigureFunc func(context.Context, *tfprotov5.ConfigureProviderRequest, *tfprotov5.ConfigureProviderResponse, *ProviderState)
+}
+
+type Resource struct {
+	Schema                *tfprotov5.Schema
+	Implementation        func(p ProviderState) ResourceImplementation
+	ImplementationContext func(p ProviderState) ResourceImplementationContext
+}
+
+// ResourceImplementation defines the methods of this resource.
+type ResourceImplementation interface {
+	Create(context.Context) []*tfprotov5.Diagnostic
+	Read(context.Context) []*tfprotov5.Diagnostic
+	Update(context.Context) []*tfprotov5.Diagnostic
+	Destroy(context.Context) []*tfprotov5.Diagnostic
+}
+```
+
+## Tradeoffs
+
+### Runtime state
+
+This section concerns stateful data generated during the lifecycle of the provider, not Terraform state.
+
+For example, consider the `meta` parameter in the CRUD functions of [`helper/schema#Resource`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema#Resource), which is conventionally used by providers to store a configured API client. Since it is sent to the provider via the `ConfigureProvider` RPC call without a way for the provider to specify a schema for this data, it cannot be strongly typed and is represented as `interface{}`. Its type must be asserted at the top of each CRUD method where it is used.
+
+#### Structs
+
+If providers, resources, and data sources are represented as structs in framework code, provider code will instantiate values of those structs. 
+
+Certain struct fields could be populated by the framework: for `Resource`, this could include `ProviderMeta` and `TerraformVersion`. Provider code cannot add additional struct fields to store strongly-typed data. This lack of flexibility is a **disadvantage**, although it likely also prevents some anti-patterns.
+
+An abandoned idea involved storing `Config` and `State` on the resource struct. This approach seems strictly worse than including config and state values in CRUD function request and response objects (e.g. `ResourceCreateRequest`), since it would be unclear when and how the framework modified those values.
+
+
+#### Interfaces

```

Comment:
One bit I think is missing from here is an acknowledgement of the disadvantage: data is strongly-typed, but that comes at the cost of needing to be _explicit_. Every bit of runtime state the framework wants to track needs to be explicitly communicated to these types, for example, `TerraformVersion`.

This can be set when the struct is created, by passing arguments to the factory:

```go
func NewComputeInstance(data framework.RuntimeState) Resource {
  return computeInstance{
    terraformVersion: data.TerraformVersion,
    client: data.ConfigureProviderData.(*Client),
  }
}
```

or it can be set with explicit methods:

```go
type computeInstanceResource struct {
  terraformVersion string
  client *Client
}

func (c *computeInstanceResource) SetTerraformVersion(version string) {
  c.terraformVersion = version
}

func (c *computeInstanceResource) SetConfigureProviderData(data framework.ConfigureProviderData) {
  c.client = data.(*Client)
}
```

It's also worth noting that helpers controlled by the framework may want access to this information (for example, tfhttp is the main audience of the TerraformVersion information, to include it in the User-Agent). This implies the necessity of a mechanism for the framework to retrieve this information. On framework-controlled structs, this is easy--it just accesses the property. On provider-controlled implementations of an interface, it needs to be accounted for with something like a getter:

```go
type computeInstanceResource struct {
  terraformVersion string
}

func (c computeInstanceResource) GetTerraformVersion() string {
  return c.terraformVersion
}
```

It may be worth calling out, though, that we could provide a framework-controlled struct providers are required to embed in their interface implementations (we can enforce this at compile time by requiring an unexported method to be implemented) that would hide this from provider developers entirely:

```go
type computeInstanceResource
  framework.RuntimeData
}
```

that would be implemented in the framework as

```go
type RuntimeData struct {
  terraformVersion string
  ConfigureProviderData interface{}
}

func (r ResourceData) getRuntimeData() framework.RuntimeData {
  return r
}
```

That would let us get access to `RuntimeData` from an `Resource` interface fulfilled by `computeInstanceResource` by just calling `getRuntimeData()`, without the provider developer needing to know anything about it, implement any methods, or do anything other than embedding the `RuntimeData` struct in their type. It would also let us hide data we wouldn't want to expose to provider developers while maintaining access to it for the framework, which might include TerraformVersion.

Reply:
Yes, it does make sense to have this included here as well as in the runtime state doc - we can remove any excessive duplication later.

Reply:
My mental model is that this document is about the tradeoff of structs and interfaces as providing structure to the framework, and the runtime data document is about the different ways we can inject that runtime data and what runtime data we want to inject, and what the tradeoffs of those options are.

Reply:
I think there's a lot of overlap here, but I think there's some context that can be teased out on either end of this, potentially with some overlap.

Reply:
It seems a bit bizarre to have `TerraformVersion` be the concern of a `Resource`, since it doesn't vary from resource to resource within the same provider. Is there any other runtime state the resource could be concerned with which the framework will need to access?

I think the `TerraformVersion` example, but with `Provider` instead of `Resource`, can be instructive here in showing what would need to happen in provider code if the framework needed access to runtime data stored on a provider-defined `Provider` struct (i.e. something fulfilling a framework-defined `Provider` interface... speaking of confusing naming). I'll write this up. 

In your final code example, what is `ResourceData`? Do you mean `Resource`?

Reply:
> It seems a bit bizarre to have `TerraformVersion` be the concern of a Resource, since it doesn't vary from resource to resource (instance) within the same provider. Is there any other runtime state the resource could be concerned with which the framework will need to access?

Agreed, but this is necessary because it [comes to us in the ConfigureProvider request](https://github.com/hashicorp/terraform-plugin-go/blob/e0e351efc90e60fa583afa6a2a2e3d8231f25e24/tfprotov6/provider.go#L140) and isn't anywhere in the [ApplyResourceChange request](https://github.com/hashicorp/terraform-plugin-go/blob/e0e351efc90e60fa583afa6a2a2e3d8231f25e24/tfprotov6/resource.go#L308-L359) meaning that if we want to include it in the User-Agent (which is why it exists), the framework is going to have to keep track of that state between the ConfigureProvider and ApplyResourceChange RPCs.

I think at the moment, that's the only framework-owned state. The API client produced by the ConfigureProvider RPC is the canonical example, but that's of a provider-defined type. But I think over time we'll continue to see information added that's framework-owned state (for example, I have proposals in mind to surface the provider address and version through ConfigureProvider RPC, etc.).

Whether that data gets embedded in a user-defined Provider type or a user-defined Resource type I think is an implementation detail that should be explored in the runtime data design doc. But the implementation doesn't materially change the way it interacts with the structs/interfaces decision being presented here.

> In your final code example, what is `ResourceData`? Do you mean `Resource`?

This is what I get for typing fast in GitHub comments. :D I mean `RuntimeData`, and have fixed it.

Reply:
I think the crux of what I'm talking about here is "how does this choice impact how we smuggle data from the ConfigureProvider RPC into the CRUD functions?" and "how does this choice impact how we can automatically enrich helpers with information from the ConfigureProvider RPC, even when the helper is called within CRUD functions?"

Reply:
Thanks, this is an extremely helpful framing and I've aimed to summarise the core points in the expanded runtime state section.